### PR TITLE
Closes #96 | Create dataset loader for LimeSoda

### LINKED ISSUE #96: Create dataset loader for LimeSoda
Dataloader name: `limesoda/limesoda.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?limesoda

|     Dataset| limesoda |
|-------------|---|
| Description | Thai fake news dataset in the healthcare domain consisting of curate and manually annotated 7,191 documents (only 4,141 documents contain token labels and are used as a test set of the baseline models). Each document in the dataset is classified as fact, fake, or undefined. |
| Subsets | - |
| Languages | tha |
| Tasks | Text Classification |
| License | Creative Commons Attribution 4.0 (cc-by-4.0) |
| Homepage | https://github.com/byinth/LimeSoda |
| HF URL | - |
| Paper URL | https://ieeexplore.ieee.org/document/9678187 |

### PR DESCRIPTION
**Pull request by jen-santoso**
Please name your PR after the issue it closes. You can use the following line: "Closes #96 " where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                        "filepath": raw_path,
+                    },
+                ),
+            ]
+        elif self.config.subset_id == "limesoda_split":
+            train_path, val_path, test_path = path_dict["split"]["train"], path_dict["split"]["val"], path_dict["split"]["test"]
+            return [
+                datasets.SplitGenerator(
+                    name=datasets.Split.TRAIN,
+                    # Whatever you put in gen_kwargs will be passed to _generate_examples
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:134]:
```suggestion
```

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    gen_kwargs={
+                        "filepath": val_path,
+                    },
+                ),
+            ]
+
+    def _generate_examples(self, filepath: Path) -> Tuple[int, Dict]:
+        with open(filepath, "r") as f:
+            df = [json.loads(line) for line in f.readlines()]
+        f.close()
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:156]: `f.close()`should be unnecessary when you're using the `with` context manager

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        if self.config.schema == "source":
+            if self.config.subset_id == "limesoda_raw":
+                for i, row in enumerate(df):
+                    ex = {"id": str(i), "title": row["Title"], "detail": row["Detail"], "title_token_tags": row["Title Token Tags"], "detail_token_tags": row["Detail Token Tags"], "document_tag": row["Document Tag"]}
+                    yield i, ex
+            else:
+                for i, row in enumerate(df):
+                    ex = {"id": str(i), "text": row["Text"], "document_tag": row["Document Tag"]}
+                    yield i, ex
+        # For example seacrowd_kb, seacrowd_t2t
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:166]:
```suggestion
```

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            description="limesoda: split SEACrowd schema",
+            schema="seacrowd_text",
+            subset_id="limesoda_split",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "limesoda_raw_source"
+
+    def _info(self) -> datasets.DatasetInfo:
+
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:89]: Running `make check_file` requires us to delete this extra newline.

```suggestion
```

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    def _split_generators(self, dl_manager: datasets.DownloadManager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+        path_dict = dl_manager.download_and_extract(_URLS)
+        if self.config.subset_id == "limesoda_raw":
+            raw_path = path_dict["raw"]
+            return [
+                datasets.SplitGenerator(
+                    name=datasets.Split.TRAIN,
+                    # Whatever you put in gen_kwargs will be passed to _generate_examples
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:123]:
```suggestion
```

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                    name=datasets.Split.VALIDATION,
+                    gen_kwargs={
+                        "filepath": val_path,
+                    },
+                ),
+            ]
+
+    def _generate_examples(self, filepath: Path) -> Tuple[int, Dict]:
+        with open(filepath, "r") as f:
+            df = [json.loads(line) for line in f.readlines()]
```
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/limesoda/limesoda.py:155]: I suggest changing the variable name `df` into something more informative. I usually see this in the context of a pandas DataFrame, but I don't see any use of that library here.

**Review by ljvmiranda921 — COMMENTED** [review]
The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename `limesoda_raw_source` into `limesoda_source`? (and `limesoda_raw_seacrowd_text` as `limesoda_seacrowd_text`). Thoughts @sabilmakbar ?

**Comment by jen-santoso** [comment]
> The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename `limesoda_raw_source` into `limesoda_source`? (and `limesoda_raw_seacrowd_text` as `limesoda_seacrowd_text`). Thoughts @sabilmakbar ?

@ljvmiranda921 Thank you for the comments. I was thinking to split the dataset into signify the one with splits and not, therefore naming the splits into `limesoda_raw_source` and `limesoda_split_source`. But think I agree with you with changing `limesoda_raw_source` to `limesoda_source`, it keeps the default runnable and it can still be easily differentiated with the splitted one.
Fixed this, and also the other comments. Please check.

**Comment by sabilmakbar** [comment]
> The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename limesoda_raw_source into limesoda_source? (and limesoda_raw_seacrowd_text as limesoda_seacrowd_text). Thoughts @sabilmakbar ?

Hi LJ @ljvmiranda921, actually for several cases, it's okay not to implement the default config of `f"{DATASET_NAME}_source"` and its SEACrowd counterpart, as long as it's justified (belongs to either Multilingual, or have different labels per config and/or has each annotator data provided; for manually curated data -- CebuaNER was one of that case). But esp for the second case, if we can be assured of one default implementation among different configs, it can be defined on `f"{DATASET_NAME}_source"`, which I agree on the fix by @jen-santoso based on ur suggestion.

> @ljvmiranda921 Thank you for the comments. I was thinking to split the dataset into signify the one with splits and not, therefore naming the splits into limesoda_raw_source and limesoda_split_source. But think I agree with you with changing limesoda_raw_source to limesoda_source, it keeps the default runnable and it can still be easily differentiated with the splitted one.
Fixed this, and also the other comments. Please check.

Ack, will verify the dataloader & its metadata in the next 24hrs. But before that, do you mind re-running your dataloader w/ this [Code Formatter Check](https://github.com/SEACrowd/seacrowd-datahub/blob/master/DATALOADER.md#5-format-your-code), if possible? I saw some lines aren't coherently formatted; I suspect the code hasn't been linted using SEACrowd Code Formatter.

**Comment by ljvmiranda921** [comment]
^Ah thanks for that reminder! I'll be more lenient on the the `_source` implementations next time.

**Review by ljvmiranda921 — APPROVED** [review]
LGTM! Let's just wait for @sabilmakbar 's comments

**Comment by jen-santoso** [comment]
> 
> Ack, will verify the dataloader & its metadata in the next 24hrs. But before that, do you mind re-running your dataloader w/ this [Code Formatter Check](https://github.com/SEACrowd/seacrowd-datahub/blob/master/DATALOADER.md#5-format-your-code), if possible? I saw some lines aren't coherently formatted; I suspect the code hasn't been linted using SEACrowd Code Formatter.

image
make check_file=seacrowd/sea_datasets/limesoda/limesoda.py result from the latest commit

**Review by sabilmakbar — CHANGES_REQUESTED** [review]
Two more nits, then all good!

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        SEACrowdConfig(
+            name="limesoda_split_seacrowd_text",
+            version=SEACROWD_VERSION,
+            description="limesoda: split SEACrowd schema",
+            schema="seacrowd_text",
+            subset_id="limesoda_split",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "limesoda_source"
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/limesoda/limesoda.py:86]: let's use `_DATASETNAME` variable
```suggestion
            name=f"{_DATASETNAME}_source",
            version=SOURCE_VERSION,
            description="limesoda source schema",
            schema="source",
            subset_id=_DATASETNAME,
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_split_source",
            version=SOURCE_VERSION,
            description="limesoda source schema",
            schema="source",
            subset_id=f"{_DATASETNAME}_split",
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_seacrowd_text",
            version=SEACROWD_VERSION,
            description="limesoda SEACrowd schema",
            schema="seacrowd_text",
            subset_id=_DATASETNAME,
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_split_seacrowd_text",
            version=SEACROWD_VERSION,
            description="limesoda: split SEACrowd schema",
            schema="seacrowd_text",
            subset_id=f"{_DATASETNAME}_split",
        ),
    ]

    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
```

**Comment by jen-santoso** [comment]
Ok, I've updated the nits suggested. Also, I added the `_LANGUAGES` constant as well to ensure consistency as recommended in https://github.com/SEACrowd/seacrowd-datahub/pull/72

**Comment by jen-santoso** [comment]
sorry I missed the class name case nits, fixed now.

File: seacrowd/sea_datasets/limesoda/limesoda.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+(only 4,141 documents contain token labels and are used as a test set of the baseline models).
+Each document in the dataset is classified as fact, fake, or undefined.
+"""
+
+_HOMEPAGE = "https://github.com/byinth/LimeSoda"
+
+_LICENSE = Licenses.CC_BY_4_0.value
+
+_LANGUAGES = ["tha"]
+
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/limesoda/limesoda.py:36]: can you add constant `_LOCAL = False`? it indicates this dataloader doesn't need an indirect download (which can't be catered by HF `DownloadManager` methods).
```suggestion
_LANGUAGES = ["tha"]
_LOCAL = False

```

**Review by sabilmakbar — COMMENTED** [review]
add `_LOCAL` variable to indicate whether this dataset needs to have indirect download or free to be downloaded w/o any gate (HF allows the dataset owner to collect user info prior to accessing dataset, which called Free upon Request dataset or Gated Dataset)

**Review by sabilmakbar — APPROVED** [review]
LGTM! Thanks @jen-santoso for implementing this dataloader and @ljvmiranda921 for reviewing this!
